### PR TITLE
feat: Setting of Routes Skeletons (REST API)(#39523)

### PR DESCRIPTION
**Pull request by ajayrose25**

### CODE REVIEW

File: src/ controllers/trainee/Controller.ts [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+
+class TraineeController {
+    static instance;
+    static getInstance = () => {
+        if (TraineeController.instance) {
+            return TraineeController.instance;
+        }
+        TraineeController.instance = new TraineeController();
+        return TraineeController.instance;
+    }
```
Comment (VisaaliSuccessive) [inline src/ controllers/trainee/Controller.ts:11]: No need of this code snippet, just export the new instance of the trainee controller while exporting

File: src/ controllers/trainee/Controller.ts [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+create = (req: Request, res: Response ) => {
```
Comment (VisaaliSuccessive) [inline src/ controllers/trainee/Controller.ts]: use next function here

File: src/ controllers/trainee/Controller.ts [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+        status: 'Ok',
+        message: 'Trainee added',
+        details: [{
+        id: 1,
+        name: 'Ajay',
+        address: 'Rohtak',
+}]
+});
+}
+list = (req: Request, res: Response ) => {
```
Comment (VisaaliSuccessive) [inline src/ controllers/trainee/Controller.ts]: Use the next function of express here

File: src/ controllers/trainee/Controller.ts [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+    }
+create = (req: Request, res: Response ) => {
+    console.log(':::::::Create Trainee:::::::');
+    res.send({
+        status: 'Ok',
+        message: 'Trainee added',
+        details: [{
+        id: 1,
+        name: 'Ajay',
+        address: 'Rohtak',
```
Comment (VisaaliSuccessive) [inline src/ controllers/trainee/Controller.ts]: Take these data in a separate object and pass it in

File: src/ controllers/trainee/Controller.ts [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+}
+list = (req: Request, res: Response ) => {
+    console.log(':::::::list Trainee:::::::');
+    res.send({
+        status: 'Ok',
+        message: 'Trainee added',
+        details: [{
+        id: 2,
+        name: 'Rahul',
+        address: 'Shahdra',
```
Comment (VisaaliSuccessive) [inline src/ controllers/trainee/Controller.ts]: take the data in a separate object and pass it in

File: src/ controllers/trainee/Controller.ts [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+}]
+});
+}
+Update = (req: Request, res: Response ) => {
```
Comment (VisaaliSuccessive) [inline src/ controllers/trainee/Controller.ts]: Use the next function of express here

File: src/ controllers/trainee/Controller.ts [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+    console.log(':::::::Update Trainee:::::::');
+    res.send({
+        status: 'Ok',
+        message: 'Trainee Updated',
+        details: [{
+        id: 3,
+        name: 'Shivam',
+        address: 'Karol Bagh'
```
Comment (VisaaliSuccessive) [inline src/ controllers/trainee/Controller.ts]: take the data in a separate object and pass it in

File: src/ controllers/trainee/Controller.ts [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+}]
+});
+}
+delete = (req: Request, res: Response ) => {
```
Comment (VisaaliSuccessive) [inline src/ controllers/trainee/Controller.ts]: Use the next function of express here

File: src/ controllers/trainee/Controller.ts [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+    console.log(':::::::Delete Trainee:::::::');
+    res.send({
+        status: 'Ok',
+        message: 'Trainee added',
+        details: [{
+        id: 4,
+        name: 'permil',
+        address: 'Rohini',
```
Comment (VisaaliSuccessive) [inline src/ controllers/trainee/Controller.ts]: take the data in a separate object and pass it in

File: src/ controllers/trainee/Controller.ts [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+        message: 'Trainee Updated',
+        details: [{
+        id: 3,
+        name: 'Shivam',
+        address: 'Karol Bagh'
+}]
+});
+}
+delete = (req: Request, res: Response ) => {
+    console.log(':::::::Delete Trainee:::::::');
```
Comment (VisaaliSuccessive) [inline src/ controllers/trainee/Controller.ts]: log the params taken

File: src/ controllers/trainee/Controller.ts [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+        message: 'Trainee added',
+        details: [{
+        id: 2,
+        name: 'Rahul',
+        address: 'Shahdra',
+}]
+});
+}
+Update = (req: Request, res: Response ) => {
+    console.log(':::::::Update Trainee:::::::');
```
Comment (VisaaliSuccessive) [inline src/ controllers/trainee/Controller.ts]: log the body taken

File: src/ controllers/trainee/Controller.ts [hunk 12]
```diff
@@ -0,0 +1,62 @@
...
+        message: 'Trainee added',
+        details: [{
+        id: 1,
+        name: 'Ajay',
+        address: 'Rohtak',
+}]
+});
+}
+list = (req: Request, res: Response ) => {
+    console.log(':::::::list Trainee:::::::');
```
Comment (VisaaliSuccessive) [inline src/ controllers/trainee/Controller.ts]: log the body taken

File: src/ controllers/trainee/Controller.ts [hunk 13]
```diff
@@ -0,0 +1,62 @@
...
+        if (TraineeController.instance) {
+            return TraineeController.instance;
+        }
+        TraineeController.instance = new TraineeController();
+        return TraineeController.instance;
+    }
+create = (req: Request, res: Response ) => {
+    console.log(':::::::Create Trainee:::::::');
+    res.send({
+        status: 'Ok',
```
Comment (VisaaliSuccessive) [inline src/ controllers/trainee/Controller.ts]: pass the status code

File: src/ controllers/trainee/Controller.ts [hunk 14]
```diff
@@ -0,0 +1,62 @@
...
+        message: 'Trainee added',
```
Comment (VisaaliSuccessive) [inline src/ controllers/trainee/Controller.ts]: Give proper user-friendly messages "Trainee Created Successfully"

File: src/ controllers/trainee/Controller.ts (hunk 13 above)
Comment (VisaaliSuccessive) [inline src/ controllers/trainee/Controller.ts]: status must be "success" not ok

File: src/ controllers/trainee/Controller.ts [hunk 15]
```diff
@@ -0,0 +1,62 @@
...
+        id: 1,
+        name: 'Ajay',
+        address: 'Rohtak',
+}]
+});
+}
+list = (req: Request, res: Response ) => {
+    console.log(':::::::list Trainee:::::::');
+    res.send({
+        status: 'Ok',
```
Comment (VisaaliSuccessive) [inline src/ controllers/trainee/Controller.ts]: status code missing and status message must be success

File: src/ controllers/trainee/Controller.ts [hunk 16]
```diff
@@ -0,0 +1,62 @@
...
+        message: 'Trainee added',
```
Comment (VisaaliSuccessive) [inline src/ controllers/trainee/Controller.ts]: incorrect message, should be - Successfully fetched trainees

File: src/ controllers/trainee/Controller.ts [hunk 17]
```diff
@@ -0,0 +1,62 @@
...
+    static getInstance = () => {
+        if (TraineeController.instance) {
+            return TraineeController.instance;
+        }
+        TraineeController.instance = new TraineeController();
+        return TraineeController.instance;
+    }
+create = (req: Request, res: Response ) => {
+    console.log(':::::::Create Trainee:::::::');
+    res.send({
```
Comment (VisaaliSuccessive) [inline src/ controllers/trainee/Controller.ts]: 1. do proper validation if not name then throw an error with 400 code and error message
2. similarly validate if not id saying id is required

File: src/ controllers/trainee/Controller.ts (hunk 11 above)
Comment (VisaaliSuccessive) [inline src/ controllers/trainee/Controller.ts]: 1. do proper validation if not id then throw an error with 400 code and error message

File: src/ controllers/trainee/Controller.ts (hunk 10 above)
Comment (VisaaliSuccessive) [inline src/ controllers/trainee/Controller.ts]: 1. do proper validation if not id then throw an error with 400 code and error message

File: src/ controllers/trainee/Controller.ts [hunk 18]
```diff
@@ -0,0 +1,62 @@
...
+        message: 'Trainee added',
+        details: [{
+        id: 4,
+        name: 'permil',
+        address: 'Rohini',
+}]
+});
+}
+}
+export default TraineeController.getInstance();
```
Comment (VisaaliSuccessive) [inline src/ controllers/trainee/Controller.ts:65]: export using new keyword

File: src/ controllers/trainee/index.ts [hunk 19]
```diff
@@ -0,0 +1,2 @@
+import traineeRouter from './routes';
+export default traineeRouter;
```
Comment (VisaaliSuccessive) [inline src/ controllers/trainee/index.ts:2]: export this in a single line using export {default} from .......

File: src/ controllers/trainee/routes.ts [hunk 20]
```diff
@@ -0,0 +1,17 @@
+import { Router } from 'express';
+import TraineeController from './Controller';
+import Controller from './Controller';
```
Comment (VisaaliSuccessive) [inline src/ controllers/trainee/routes.ts:3]: Unnecessary import statement

File: src/ controllers/trainee/routes.ts [hunk 21]
```diff
@@ -0,0 +1,17 @@
+import { Router } from 'express';
+import TraineeController from './Controller';
```
Comment (VisaaliSuccessive) [inline src/ controllers/trainee/routes.ts:2]: import only the trainee module, not the controller

File: src/ controllers/trainee/routes.ts [hunk 22]
```diff
@@ -0,0 +1,17 @@
...
+import Controller from './Controller';
+
+const traineeRouter = Router();
```
Comment (VisaaliSuccessive) [inline src/ controllers/trainee/routes.ts:5]: Define interface for the router of type router (express)

File: src/ controllers/trainee/routes.ts [hunk 23]
```diff
@@ -0,0 +1,17 @@
...
+
+traineeRouter.route('/')
+.get(TraineeController.create)
+.post(TraineeController.list)
+.delete(TraineeController.delete)
+.put(TraineeController.Update);
```
Comment (VisaaliSuccessive) [inline src/ controllers/trainee/routes.ts:11]: please check this, it is not done in right way. 
get isfor listing 
post is for creating 
put is for updating 
delete for deleting 
Please go through these concepts properly

traineeRouter
    .get('/', trainee.get) 
    .post('/', trainee.create)
    .put('/', trainee.update)
    .delete('/:id', trainee.delete);

File: src/Server.ts [hunk 24]
```diff
@@ -2,6 +2,7 @@ import * as express from 'express';
 import * as bodyParser from 'body-parser';
 import Iconfig from './config/iconfig';
 import { errorHandler, notFoundRoute } from './libs/routes';
+import router from './router';
 import { eventNames } from 'cluster';
```
Comment (VisaaliSuccessive) [inline src/Server.ts:6]: remove unwanted imports

File: src/router.ts [hunk 25]
```diff
@@ -0,0 +1,8 @@
+import  traineeRouter  from './ controllers/trainee/index';
```
Comment (VisaaliSuccessive) [inline src/router.ts]: remove index from import statement, please make sure not to repeat mistakes

File: src/router.ts [hunk 26]
```diff
@@ -0,0 +1,8 @@
...
+import { Router } from 'express';
+
```
Comment (VisaaliSuccessive) [inline src/router.ts:3]: where is the importing of user route?

File: src/router.ts [hunk 27]
```diff
@@ -0,0 +1,8 @@
...
+const mainRouter = Router();
```
Comment (VisaaliSuccessive) [inline src/router.ts]: define the interface of express route type

File: src/router.ts [hunk 28]
```diff
@@ -0,0 +1,8 @@
...
+mainRouter.use('/trainee', traineeRouter);
+
```
Comment (VisaaliSuccessive) [inline src/router.ts:6]: where is user route?

File: src/router.ts [hunk 29]
```diff
@@ -0,0 +1,8 @@
...
+export default mainRouter;
```
Comment (VisaaliSuccessive) [inline src/router.ts]: rename this as per the name of the file "router"

**Review by VisaaliSuccessive — COMMENTED** [review]
user module is not defined similar to trainee

File: src/ controllers/trainee/Controller.ts (hunk 1 above)
Comment (ajayrose25) [reply src/ controllers/trainee/Controller.ts:11]: Done

File: src/ controllers/trainee/Controller.ts (hunk 2 above)
Comment (ajayrose25) [reply src/ controllers/trainee/Controller.ts]: Done

File: src/ controllers/trainee/Controller.ts (hunk 3 above)
Comment (ajayrose25) [reply src/ controllers/trainee/Controller.ts]: Done

File: src/ controllers/trainee/Controller.ts (hunk 4 above)
Comment (ajayrose25) [reply src/ controllers/trainee/Controller.ts]: Done

File: src/ controllers/trainee/Controller.ts (hunk 5 above)
Comment (ajayrose25) [reply src/ controllers/trainee/Controller.ts]: Done

File: src/router.ts (hunk 25 above)
Comment (ajayrose25) [reply src/router.ts]: Done

File: src/router.ts (hunk 26 above)
Comment (ajayrose25) [reply src/router.ts:3]: Done

File: src/router.ts (hunk 27 above)
Comment (ajayrose25) [reply src/router.ts]: Done

File: src/router.ts (hunk 28 above)
Comment (ajayrose25) [reply src/router.ts:6]: Done

File: src/router.ts (hunk 29 above)
Comment (ajayrose25) [reply src/router.ts]: Done

**Review by VisaaliSuccessive — APPROVED** [review]

**Pull request merged**